`invoke local` is not supported

Hi. New to Serverless so maybe I'm doing it wrong, but:

### Steps
 * Check out and build example project
 * Verify that `sls invoke hello -d "[1, 100, 10]"` correctly returns `"Highest value is 100"`
 * Then run `sls invoke local hello -d "[1, 100, 10]"`

### Expected
 * Same result, i.e. `"Highest value is 100"`

### Actual
Errors with:
```
{ Error: Cannot find module '/home/nick/workspace/GITHUB/serverless-haskell/example-project/serverless-haskell-example'                                                            
    at Function.Module._resolveFilename (module.js:547:15)                               
    at Function.Module._load (module.js:474:25)                                          
    at Module.require (module.js:596:17)    
    at require (internal/module.js:11:18)   
    at AwsInvokeLocal.invokeLocalNodeJs (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:258:33)       
    at AwsInvokeLocal.invokeLocal (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/serverless/lib/plugins/aws/invokeLocal/index.js:125:19)             
    at AwsInvokeLocal.tryCatcher (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/util.js:16:23)                                   
    at Promise._settlePromiseFromHandler (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/promise.js:512:31)                       
    at Promise._settlePromise (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/promise.js:569:18)                                  
    at Promise._settlePromise0 (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/promise.js:614:10)                                 
    at Promise._settlePromises (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/promise.js:693:18)                                 
    at Async._drainQueue (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/async.js:133:16)                                         
    at Async._drainQueues (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/async.js:143:10)                                        
    at Immediate.Async.drainQueues (/home/nick/workspace/GITHUB/serverless-haskell/example-project/node_modules/bluebird/js/release/async.js:17:14)                                
    at runCallback (timers.js:794:20)       
    at tryOnImmediate (timers.js:752:5)     
    at processImmediate [as _immediateCallback] (timers.js:729:5) code: 'MODULE_NOT_FOUND' }                                                                                       
Done in 0.66s.                              
```

--- a/README.md
+++ b/README.md
@@ -83,6 +83,8 @@ Deploying Haskell code onto [AWS Lambda] using [Serverless].
   You can test the function and see the invocation results with `sls invoke
   myfunc`.
 
+  To invoke the function locally, use `sls invoke local -f myfunc`.
+
 ### Notes
 
 * `sls deploy function` is [not supported yet](https://github.com/seek-oss/serverless-haskell/issues/20).

--- a/README.md
+++ b/README.md
@@ -85,6 +85,15 @@ Deploying Haskell code onto [AWS Lambda] using [Serverless].
 
   To invoke the function locally, use `sls invoke local -f myfunc`.
 
+### API Gateway
+
+This plugin supports handling API Gateway requests. Declare the HTTP events
+normally in `serverless.yml` and use
+[AWSLambda.Events.APIGateway](https://hackage.haskell.org/package/serverless-haskell/docs/AWSLambda-Events-APIGateway.html)
+in the handler to process them.
+
+[Serverless Offline] can be used for local testing of API Gateway requests.
+
 ### Notes
 
 * `sls deploy function` is [not supported yet](https://github.com/seek-oss/serverless-haskell/issues/20).
@@ -129,4 +138,5 @@ series, use `RESOLVER_SERIES=lts-9`.
 [jq]: https://stedolan.github.io/jq/
 [NPM]: https://www.npmjs.com/
 [Serverless]: https://serverless.com/framework/
+[Serverless Offline]: https://github.com/dherault/serverless-offline
 [Stack]: https://haskellstack.org

--- a/README.md
+++ b/README.md
@@ -86,7 +86,6 @@ Deploying Haskell code onto [AWS Lambda] using [Serverless].
 ### Notes
 
 * `sls deploy function` is [not supported yet](https://github.com/seek-oss/serverless-haskell/issues/20).
-* `sls invoke local` is [not supported yet](https://github.com/seek-oss/serverless-haskell/issues/32).
 * Only AWS Lambda is supported at the moment. Other cloud providers would
   require different JavaScript wrappers to be implemented.
 

--- a/example-project/package.json
+++ b/example-project/package.json
@@ -6,7 +6,7 @@
     "test": "test"
   },
   "dependencies": {
-    "serverless": "^1.26.0",
+    "serverless": "^1.27.3",
     "serverless-haskell": "file:../serverless-plugin"
   },
   "devDependencies": {},

--- a/integration-test/expected/local_output.txt
+++ b/integration-test/expected/local_output.txt
@@ -1,0 +1,9 @@
+This should go to logs
+["one","two","--three"]
+Just ["abc"]
+Array [Number 4.0,Number 5.0,Number 6.0]
+[
+    1,
+    2,
+    3
+]

--- a/integration-test/expected/offline_output.txt
+++ b/integration-test/expected/offline_output.txt
@@ -1,0 +1,1 @@
+Hello, integration

--- a/integration-test/run.sh
+++ b/integration-test/run.sh
@@ -80,20 +80,32 @@ done
 
 export PATH=$(npm bin):$PATH
 
-# Install Serverless and deploy the project
+# Install Serverless
 npm install serverless
 npm install $DIST/serverless-plugin
+npm install serverless-offline
 
+# Just package the service first
+sls package
+echo "Packaging verified."
+
+# Test local invocation
 sls invoke local --function main --data '[4, 5, 6]' | \
     grep -v 'Serverless: ' > local_output.txt
 
 diff $EXPECTED/local_output.txt local_output.txt && echo "Expected local result verified."
 
+# Test serverless-offline
+sls offline start --exec \
+    "sh -c 'curl -s http://localhost:3000/hello/integration > offline_output.txt'"
+diff $EXPECTED/offline_output.txt offline_output.txt && echo "Expected serverless-offline result verified."
+
 if [ "$DRY_RUN" = "true" ]
 then
-    sls package
-    echo "Packaging verified."
+    # All done (locally)
+    exit 0
 else
+    # Deploy to AWS
     sls deploy
 
     # Run the function and verify the results

--- a/integration-test/run.sh
+++ b/integration-test/run.sh
@@ -84,6 +84,11 @@ export PATH=$(npm bin):$PATH
 npm install serverless
 npm install $DIST/serverless-plugin
 
+sls invoke local --function main --data '[4, 5, 6]' | \
+    grep -v 'Serverless: ' > local_output.txt
+
+diff $EXPECTED/local_output.txt local_output.txt && echo "Expected local result verified."
+
 if [ "$DRY_RUN" = "true" ]
 then
     sls package

--- a/integration-test/skeleton/ApiGateway.hs
+++ b/integration-test/skeleton/ApiGateway.hs
@@ -1,0 +1,19 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+module Main where
+
+import AWSLambda.Events.APIGateway
+import Control.Lens
+import qualified Data.HashMap.Strict as HashMap
+import Data.Semigroup
+import Data.Text (Text)
+
+main :: IO ()
+main = apiGatewayMain hello
+
+hello :: APIGatewayProxyRequest Text -> IO (APIGatewayProxyResponse Text)
+hello request = do
+  putStrLn "This should go to logs"
+  case HashMap.lookup "name" (request ^. agprqPathParameters) of
+    Just name -> return $ responseOK & responseBody ?~ "Hello, " <> name <> "\n"
+    Nothing -> return responseNotFound

--- a/integration-test/skeleton/package.yaml
+++ b/integration-test/skeleton/package.yaml
@@ -14,3 +14,14 @@ executables:
     - -threaded
     - -rtsopts
     - -with-rtsopts=-N
+
+  apigw:
+    main: ApiGateway.hs
+    ghc-options:
+    - -threaded
+    - -rtsopts
+    - -with-rtsopts=-N
+    dependencies:
+      - lens
+      - text
+      - unordered-containers

--- a/integration-test/skeleton/serverless.yml
+++ b/integration-test/skeleton/serverless.yml
@@ -8,11 +8,19 @@ functions:
   main:
     handler: shtest.main
 
+  apigw:
+    handler: shtest.apigw
+    events:
+      - http:
+          path: hello/{name}
+          method: get
+
   subdir:
     handler: subdir/subtest.main
 
 plugins:
 - serverless-haskell
+- serverless-offline
 
 custom:
   haskell:

--- a/serverless-plugin/index.js
+++ b/serverless-plugin/index.js
@@ -178,7 +178,7 @@ class ServerlessPlugin {
     buildHandlersLocal(options) {
         options = options || {};
         this.buildHandlers(Object.assign(options, {
-            noDocker: true
+            localRun: true
         }));
     }
 
@@ -186,7 +186,7 @@ class ServerlessPlugin {
         const service = this.serverless.service;
 
         options = options || {};
-        if (options.noDocker) {
+        if (options.localRun) {
             this.docker.required = false;
         }
 
@@ -239,31 +239,33 @@ class ServerlessPlugin {
             this.additionalFiles.push(targetPath);
             this.addToHandlerOptions(handlerOptions, funcName, targetDirectory, packageName, executableName);
 
-            // Copy libraries not present on AWS Lambda environment
-            const lddOutput = this.runStackOutput(
-                directory,
-                [
-                    'exec',
-                    'ldd',
-                    executablePath,
-                ]
-            );
-            const executableLibraries = ld.parseLdOutput(lddOutput);
+            if (!options.localRun) {
+                // Copy libraries not present on AWS Lambda environment
+                const lddOutput = this.runStackOutput(
+                    directory,
+                    [
+                        'exec',
+                        'ldd',
+                        executablePath,
+                    ]
+                );
+                const executableLibraries = ld.parseLdOutput(lddOutput);
 
-            for (const name in executableLibraries) {
-                if (!libraries[name] && !IGNORE_LIBRARIES.includes(name)) {
-                    const libPath = executableLibraries[name];
-                    const targetPath = path.resolve(this.servicePath, name);
-                    this.runStack(
-                        directory,
-                        [
-                            'exec',
-                            'cp',
-                            libPath,
-                            targetPath,
-                        ]);
-                    this.additionalFiles.push(targetPath);
-                    libraries[name] = true;
+                for (const name in executableLibraries) {
+                    if (!libraries[name] && !IGNORE_LIBRARIES.includes(name)) {
+                        const libPath = executableLibraries[name];
+                        const targetPath = path.resolve(this.servicePath, name);
+                        this.runStack(
+                            directory,
+                            [
+                                'exec',
+                                'cp',
+                                libPath,
+                                targetPath,
+                            ]);
+                        this.additionalFiles.push(targetPath);
+                        libraries[name] = true;
+                    }
                 }
             }
         });

--- a/serverless-plugin/index.js
+++ b/serverless-plugin/index.js
@@ -34,6 +34,8 @@ class ServerlessPlugin {
         this.hooks = {
             'before:package:createDeploymentArtifacts': this.beforeCreateDeploymentArtifacts.bind(this),
             'after:package:createDeploymentArtifacts': this.afterCreateDeploymentArtifacts.bind(this),
+            // serverless-offline
+            'before:offline:start:init': this.beforeCreateDeploymentArtifacts.bind(this),
         };
 
         this.servicePath = this.serverless.config.servicePath || '';

--- a/serverless-plugin/index.js
+++ b/serverless-plugin/index.js
@@ -34,6 +34,10 @@ class ServerlessPlugin {
         this.hooks = {
             'before:package:createDeploymentArtifacts': this.beforeCreateDeploymentArtifacts.bind(this),
             'after:package:createDeploymentArtifacts': this.afterCreateDeploymentArtifacts.bind(this),
+
+            // invoke local
+            'before:invoke:local:invoke': this.beforeCreateDeploymentArtifacts.bind(this),
+            'after:invoke:local:invoke': this.afterCreateDeploymentArtifacts.bind(this),
             // serverless-offline
             'before:offline:start:init': this.beforeCreateDeploymentArtifacts.bind(this),
         };

--- a/serverless-plugin/index.js
+++ b/serverless-plugin/index.js
@@ -32,14 +32,16 @@ class ServerlessPlugin {
         this.options = options;
 
         this.hooks = {
-            'before:package:createDeploymentArtifacts': this.beforeCreateDeploymentArtifacts.bind(this),
-            'after:package:createDeploymentArtifacts': this.afterCreateDeploymentArtifacts.bind(this),
+            'before:package:createDeploymentArtifacts': this.buildHandlers.bind(this),
+            'after:package:createDeploymentArtifacts': this.cleanupHandlers.bind(this),
 
             // invoke local
-            'before:invoke:local:invoke': this.beforeCreateDeploymentArtifacts.bind(this),
-            'after:invoke:local:invoke': this.afterCreateDeploymentArtifacts.bind(this),
+            'before:invoke:local:invoke': this.buildHandlersLocal.bind(this),
+            'after:invoke:local:invoke': this.cleanupHandlers.bind(this),
+
             // serverless-offline
-            'before:offline:start:init': this.beforeCreateDeploymentArtifacts.bind(this),
+            'before:offline:start:init': this.buildHandlersLocal.bind(this),
+            'after:offline:start:end': this.cleanupHandlers.bind(this),
         };
 
         this.servicePath = this.serverless.config.servicePath || '';
@@ -173,8 +175,20 @@ class ServerlessPlugin {
         }]);
     }
 
-    beforeCreateDeploymentArtifacts() {
+    buildHandlersLocal(options) {
+        options = options || {};
+        this.buildHandlers(Object.assign(options, {
+            noDocker: true
+        }));
+    }
+
+    buildHandlers(options) {
         const service = this.serverless.service;
+
+        options = options || {};
+        if (options.noDocker) {
+            this.docker.required = false;
+        }
 
         // Exclude Haskell artifacts from uploading
         service.package.exclude = service.package.exclude || [];
@@ -257,8 +271,8 @@ class ServerlessPlugin {
         this.writeHandlers(handlerOptions);
     }
 
-    afterCreateDeploymentArtifacts() {
-       this.additionalFiles.forEach(fileName => fs.removeSync(fileName));
+    cleanupHandlers(options) {
+        this.additionalFiles.forEach(fileName => fs.removeSync(fileName));
     }
 }
 

--- a/serverless-plugin/package.json
+++ b/serverless-plugin/package.json
@@ -16,7 +16,7 @@
   "dependencies": {
     "fs-copy-file-sync": "^1.1.1",
     "fs-extra": "^5.0.0",
-    "serverless": "^1.27.2"
+    "serverless": "^1.27.3"
   },
   "devDependencies": {
     "eslint": "^4.19.1"

--- a/src/AWSLambda.hs
+++ b/src/AWSLambda.hs
@@ -1,5 +1,5 @@
 {-|
-Module      : AWSLambda.Handler
+Module      : AWSLambda
 Stability   : experimental
 Portability : POSIX
 
@@ -46,6 +46,15 @@ To deploy a Haskell function on AWS Lambda:
   myfunc@.
 
   To invoke the function locally, use @sls invoke local -f myfunc@.
+
+= API Gateway
+
+This plugin supports handling API Gateway requests. Declare the HTTP events
+normally in @serverless.yml@ and use 'AWSLambda.Events.APIGateway' in the
+handler to process them.
+
+<https://github.com/dherault/serverless-offline Serverless Offline> can be used
+for local testing of API Gateway requests.
 
 = Additional features
 

--- a/src/AWSLambda.hs
+++ b/src/AWSLambda.hs
@@ -45,6 +45,8 @@ To deploy a Haskell function on AWS Lambda:
   You can test the function and see the invocation results with @sls invoke
   myfunc@.
 
+  To invoke the function locally, use @sls invoke local -f myfunc@.
+
 = Additional features
 
 Configuration options are passed to the plugin under @haskell@ key in @custom@

--- a/src/AWSLambda/Events/APIGateway.hs
+++ b/src/AWSLambda/Events/APIGateway.hs
@@ -8,6 +8,19 @@ Module: AWSLambda.Events.APIGateway
 Description: Types for APIGateway Lambda requests and responses
 
 Based on https://github.com/aws/aws-lambda-dotnet/tree/master/Libraries/src/Amazon.Lambda.APIGatewayEvents
+
+To enable processing of API Gateway events, use the @events@ key in
+@serverless.yml@ as usual:
+
+> functions:
+>   myapifunc:
+>     handler: mypackage.mypackage-exe
+>     events:
+>       - http:
+>           path: hello/{name}
+>           method: get
+
+Then use 'apiGatewayMain' in the handler to process the requests.
 -}
 module AWSLambda.Events.APIGateway where
 
@@ -200,36 +213,36 @@ responseBodyEmbedded = responseBody . mapping unEmbed
 responseBodyBinary :: Setter' (APIGatewayProxyResponse Base64) (Maybe ByteString)
 responseBodyBinary = responseBody . mapping _Base64
 
--- | Process incoming events from @serverless-haskell@ using a provided
--- function.
---
--- This is a specialisation of lambdaMain for API Gateway.
---
--- The handler receives the input event given to the AWS Lambda function, and
--- its return value is returned from the function.
---
--- This is intended to be used as @main@, for example:
---
--- > import AWSLambda.Events.APIGateway
--- > import Control.Lens
--- > import Data.Aeson
--- > import Data.Aeson.Embedded
--- >
--- > main = apiGatewayMain handler
--- >
--- > handler :: APIGatewayProxyRequest (Embedded Value) -> IO (APIGatewayProxyResponse (Embedded [Int]))
--- > handler request = do
--- >   putStrLn "This should go to logs"
--- >   print $ request ^. requestBody
--- >   pure $ responseOK & responseBodyEmbedded ?~ [1, 2, 3]
---
--- The type parameters @reqBody@ and @resBody@ represent the types of request and response body, respectively.
--- The @FromText@ and @ToText@ contraints are required because these values come from string fields
--- in the request and response JSON objects.
--- To get direct access to the body string, use @Text@ as the parameter type.
--- To treat the body as a stringified embedded JSON value, use @Embedded a@, where @a@ has the
--- appropriate @FromJSON@ or @ToJSON@ instances.
--- To treat the body as base 64 encoded binary use @Base64@.
+{-| Process incoming events from @serverless-haskell@ using a provided function.
+
+This is a specialisation of 'lambdaMain' for API Gateway.
+
+The handler receives the input event given to the AWS Lambda function, and
+its return value is returned from the function.
+
+This is intended to be used as @main@, for example:
+
+> import AWSLambda.Events.APIGateway
+> import Control.Lens
+> import Data.Aeson
+> import Data.Aeson.Embedded
+>
+> main = apiGatewayMain handler
+>
+> handler :: APIGatewayProxyRequest (Embedded Value) -> IO (APIGatewayProxyResponse (Embedded [Int]))
+> handler request = do
+>   putStrLn "This should go to logs"
+>   print $ request ^. requestBody
+>   pure $ responseOK & responseBodyEmbedded ?~ [1, 2, 3]
+
+The type parameters @reqBody@ and @resBody@ represent the types of request and response body, respectively.
+The @FromText@ and @ToText@ contraints are required because these values come from string fields
+in the request and response JSON objects.
+To get direct access to the body string, use @Text@ as the parameter type.
+To treat the body as a stringified embedded JSON value, use @Embedded a@, where @a@ has the
+appropriate @FromJSON@ or @ToJSON@ instances.
+To treat the body as base 64 encoded binary use @Base64@.
+-}
 apiGatewayMain
   :: (FromText reqBody, ToText resBody)
   => (APIGatewayProxyRequest reqBody -> IO (APIGatewayProxyResponse resBody)) -- ^ Function to process the event

--- a/src/AWSLambda/Handler.hs
+++ b/src/AWSLambda/Handler.hs
@@ -6,6 +6,7 @@ Portability : POSIX
 Entry point for AWS Lambda handlers deployed with @serverless-haskell@ plugin.
 -}
 {-# LANGUAGE TypeApplications #-}
+
 module AWSLambda.Handler
   ( lambdaMain
   ) where
@@ -19,11 +20,12 @@ import qualified Data.ByteString as ByteString
 
 import qualified Data.Text.Lazy.IO as Text
 
-import GHC.IO.Handle (Handle, hClose)
+import GHC.IO.Handle
+       (BufferMode(..), Handle, hClose, hSetBuffering)
 
 import System.IO (stdout)
-import System.Posix.IO (fdToHandle)
 import System.Posix.Files (getFdStatus)
+import System.Posix.IO (fdToHandle)
 import System.Posix.Types (Fd(..))
 
 -- | Process incoming events from @serverless-haskell@ using a provided
@@ -89,14 +91,17 @@ lambdaMain act =
         Text.hPutStrLn resultChannel $ Aeson.encodeToLazyText result
         pure ()
 
--- | Invoke an action with the handle to write the results to. On AWS Lambda,
--- use the channel opened by the JavaScript wrapper, otherwise use standard
--- output.
+-- | Invoke an action with the handle to write the results to. If called by the
+-- JavaScript wrapper, use the channel opened by it, otherwise use standard
+-- output. Also set line buffering on standard output for AWS Lambda so the logs
+-- are output in a timely manner.
 withResultChannel :: (Handle -> IO r) -> IO r
 withResultChannel act = do
   commStatus <- try @IOError $ getFdStatus communicationFd
   case commStatus of
-    Right _ -> bracket (fdToHandle communicationFd) hClose act
+    Right _ -> do
+      hSetBuffering stdout LineBuffering
+      bracket (fdToHandle communicationFd) hClose act
     Left _ -> act stdout
 
 -- | File descriptor opened by the JavaScript wrapper to listen for the results

--- a/src/AWSLambda/Handler.hs
+++ b/src/AWSLambda/Handler.hs
@@ -5,11 +5,12 @@ Portability : POSIX
 
 Entry point for AWS Lambda handlers deployed with @serverless-haskell@ plugin.
 -}
+{-# LANGUAGE TypeApplications #-}
 module AWSLambda.Handler
   ( lambdaMain
   ) where
 
-import Control.Exception (bracket)
+import Control.Exception (bracket, try)
 
 import qualified Data.Aeson as Aeson
 import qualified Data.Aeson.Text as Aeson
@@ -20,9 +21,9 @@ import qualified Data.Text.Lazy.IO as Text
 
 import GHC.IO.Handle (Handle, hClose)
 
-import System.Environment (lookupEnv)
 import System.IO (stdout)
 import System.Posix.IO (fdToHandle)
+import System.Posix.Files (getFdStatus)
 import System.Posix.Types (Fd(..))
 
 -- | Process incoming events from @serverless-haskell@ using a provided
@@ -93,10 +94,10 @@ lambdaMain act =
 -- output.
 withResultChannel :: (Handle -> IO r) -> IO r
 withResultChannel act = do
-  lambdaFunctionName <- lookupEnv "AWS_LAMBDA_FUNCTION_NAME"
-  case lambdaFunctionName of
-    Just _ -> bracket (fdToHandle communicationFd) hClose act
-    Nothing -> act stdout
+  commStatus <- try @IOError $ getFdStatus communicationFd
+  case commStatus of
+    Right _ -> bracket (fdToHandle communicationFd) hClose act
+    Left _ -> act stdout
 
 -- | File descriptor opened by the JavaScript wrapper to listen for the results
 communicationFd :: Fd

